### PR TITLE
Created valid_move? method for knight, complete with tests for said m…

### PR DESCRIPTION
**Pull request by rjdellecese**
…ethod.

@lasley Did I go overboard at all with the tests here [serious]? Other than that, I think it's good to go!

### CODE REVIEW

**Comment by rjdellecese** [comment]
@lasley Actually, one thing to note here is that I'll need to incorporate the piece of the `is_obstructed?` method that checks whether the destination square is occupied by a piece of the same color. Ideally, the `is_obstructed?` method would itself call a different method within it (`is_occupied?`, perhaps) that I could just call on the knight myself (without using `super`). Should I stub something like that in the knight method? Or a test that accounts for that outcome?

**Comment by Mikeysax** [comment]
Hey @rjdellecese I'm actually added this into the capture function so possibly wait for my merge?

File: app/models/knight.rb [hunk 1]
```diff
@@ -1,2 +1,7 @@
 class Knight < ChessPiece
+  def valid_move?(next_x, next_y)
+    return false if out_of_bounds?(next_x, next_y)
+    return false unless x_diff(next_x) == 2 && y_diff(next_y) == 1 || x_diff(next_x) == 1 && y_diff(next_y) == 2
```
Comment (lasley) [inline app/models/knight.rb]: Maybe split this into two conditions for clarity, this is a pretty long string of statements.

File: spec/models/knight_spec.rb [hunk 2]
```diff
@@ -1,5 +1,107 @@
 require 'rails_helper'
 
 RSpec.describe Knight, type: :model do
-  pending "add some examples to (or delete) #{__FILE__}"
+  describe 'valid_move? for knight on b2 (position_x: 2, position_y: 2)' do
```
Comment (lasley) [inline spec/models/knight_spec.rb]: Typically you want to try to keep to one expect/assert per test. This is so that one isolated edge case doesn't break the entire set of tests, which in turn will yield a clearer picture of where the failure actually is (think 2 broken tests, 2 working - instead of one broken).

Instead of just testing everything, try to think of the edge cases that could be caused by your code, then test those independently. For valid_move, I see the following:
* False if out_of_bounds? returns true (don't actually test anything in out_of_bounds?, you only need setup the condition to make it false so that we meet the condition of the logic)
* False unless x_diff(next_x) == 2 && y_diff(next_y) == 1
* False unless x_diff(next_x) == 1 && y_diff(next_y) == 2
* True if anything else

File: app/models/knight.rb (hunk 1 above)
Comment (Mikeysax) [inline app/models/knight.rb]: out_of_bounds? is already on the super method.

Comment (lasley) [inline app/models/knight.rb]: Good catch, we would want to `return false if not super(next_x, next_y)`

File: spec/models/knight_spec.rb (hunk 2 above)
Comment (Mikeysax) [inline spec/models/knight_spec.rb]:
```class King < ChessPiece
  def valid_move?(next_x, next_y)
    return false unless within_movement_constraints(next_x, next_y)
    super(next_x, next_y)
  end

  private

  def within_movement_constraints(next_x, next_y)
    x_diff(next_x) == 1 || y_diff(next_y) == 1
  end
end
```
I've split the logic into a private method within the king model. Maybe try this?

File: app/models/knight.rb (hunk 1 above)
Comment (Mikeysax) [inline app/models/knight.rb]:
```
def valid_move?(next_x, next_y)
    return false unless within_movement_constraints(next_x, next_y)
    super(next_x, next_y)
  end
```

Comment (rjdellecese) [inline app/models/knight.rb]: @lasley @Mikeysax I didn't include `super` because `valid_move?` for any old chess piece will check for obstruction, which I don't want to occur for the knight. I couldn't find a way to call `super` but exclude some part of the inherited method, so I opted to instead just call the parts of the chess piece model's `valid_move?` method that are relevant to the knight (ie. everything else)

Comment (rjdellecese) [inline app/models/knight.rb]: Actually, upon further reflection, I think the best thing would probably be to place a conditional in the `obstructed?` or `valid_move?` method that acts differently if the piece.type is knight. That way I can still call `super` and I won't have to worry about having to copy any new logic in the `valid_move?` knight method should anything in the `valid_move?` chess piece method change (as it will, I think, in future tasks).

Comment (Mikeysax) [inline app/models/knight.rb]: Yeah, that's prob. better if we do that.

**Comment by lasley** [comment]
@rjdellecese - If you wanna knock out the merge conflicts & the failing test, then this is :+1:

**Comment by rjdellecese** [comment]
@lasley I think this is ready to gooo

**Comment by lasley** [comment]
:+1:
